### ISSUE TITLE
SINV is more expensive than ASND/ARCV

### ISSUE THREAD

**Issue by phanikishoreg** [gparmer/composite#236]
On baremetal (CPU Speed:2.8GHz), 
* SINV Avg: 391, WCET = 536, 
* ASND/ARCV Avg: 350, WCET: 658. 

Average on SINV is slower than ASND. Mis-optimization somewhere, need to look into that.

**Comment by interwq** [comment]
I'm pretty curious here: are the numbers round trip or one-way? Does the ASND/ARCV cost involve context switch cost like SINV does, or in other words how it's measured?

**Comment by gparmer** [comment]
sinv = from sender to receiver (one way) -- we have the separate measurement for sret (349 cycles).

asnd/rcv is rdtsc in sender, and, assuming an immediate switch, rdtsc after rcv.  But your question is good:  do we write to cr3 in this case?  I actually do *not* think that we do.  In that case, these numbers are expected.  @phanikishoreg confirmation?

@interwq we're finally writing up the tcap stuff so we're doing proper prioritized dispatch on interrupts and asnd.  We have a Xen-like system where VMs pass tcap time with function requests for I/O for Dom0.  We can send you the paper after we're done if you're interested.

**Comment by phanikishoreg** [comment]
No, we don't write to cr3. These numbers are for sinv and asnd/arcv operations within a component.

**Comment by interwq** [comment]
Ah I see. Thanks for clarifying @gparmer and @phanikishoreg. Yeah if that's the case, SINV costs more isn't crazy -- IIRC updating cr3 accounts for a fair chunk of the SINV path.

I'm definitely be interested in the tcap paper. Proper prioritization sounds great :)
